x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/interceptor/BulkShardRequestInterceptor.java

```diff
@@ -41,7 +42,8 @@ public BulkShardRequestInterceptor(ThreadPool threadPool, XPackLicenseState lice
     @Override
     public void intercept(RequestInfo requestInfo, AuthorizationEngine authzEngine, AuthorizationInfo authorizationInfo,
                           ActionListener<Void> listener) {
-        boolean shouldIntercept = licenseState.isSecurityEnabled() && licenseState.checkFeature(Feature.SECURITY_DLS_FLS);
+        boolean shouldIntercept = licenseState.isSecurityEnabled();
+        var licenseChecker = new MemoizedSupplier<>(() -> licenseState.checkFeature(Feature.SECURITY_DLS_FLS));
```

Comment:
The existing expression of `licenseState.isSecurityEnabled() && licenseState.checkFeature(Feature.SECURITY_DLS_FLS)` has a potential racing issue between the two license checks. I noticed we have been somewhat inconsistent in handling this. In some places, e.g. `ResizeRequestInterceptor`, we use `licenseState.copyCurrentLicenseState()` to ensure checks are always done against the same license. In other places including this one, no special handling is done. The new changes also retain this behaviour. My question is: what is our recommend approach for this subtle issue? I am OK with either approach as long as it is consistently applied.

I am also curious about the benefit of the new `MemoizedSupplier`. Why do we need to check the license deeper in the loop? Isn't it sufficient and more efficient to check in the outermost `if` statement?

Reply:
> what is our recommend approach for this subtle issue?

It is true there is a race condition, but in practice I don't think we need to worry about it. This is an extreme edge case, where the license changes mid request. AFAIU, the copy method was added to handle the race condition in tests, not for any problem found by users. That said, the long term solution to this is to change all of our checks to use a stashed copy in the ThreadContext that would be inserted when the request is first received. That idea is described in https://github.com/elastic/elasticsearch/issues/53909.

> I am also curious about the benefit of the new MemoizedSupplier. Why do we need to check the license deeper in the loop?

We don't know if DLS/FLS is actually used until we resolve the each indices' access controls inside the loop. Thus we need to check when resolving each index. It is true this is similar to the problem described above, and could be partially solved by copying the license state. However, it also means needlessly updating the last used time in quick succession. I could go either way, but thought reducing those updates was worthwhile given the simpleness of implementing the MemoizedSupplier.